src/coverage_reporter/parsers/coveragepy_parser.cr

```diff
@@ -30,114 +27,22 @@ module CoverageReporter
       false
     end
 
-    def parse(filename : String) : Array(FileReport)
-      lines = {} of String => Array(Hits)
-
-      DB.open "sqlite3://#{filename}" do |db|
-        db.query(QUERY) do |rs|
-          rs.each do
-            name = rs.read(String)
-            numbits = rs.read(Slice(UInt8))
-            nums = [] of Hits
-            numbits.each_with_index do |byte, byte_i|
-              8.times do |bit_i|
-                if byte & (1 << bit_i) != 0
-                  nums << (byte_i * 8 + bit_i).to_u64
-                end
-              end
-            end
-            lines[name] = nums
-          end
-        end
-      end
-
-      lines.map do |name, hits|
-        coverage = get_coverage(name, hits)
-
-        file_report(
-          name: name,
-          coverage: coverage,
-        )
-      end
-    end
-
-    private def get_coverage(name : String, hits : Array(Hits)) : Array(Hits?)
-      coverage = {} of Line => Hits?
-
-      line_no = 1.to_u64
-      under_def = false
-      docstring = false
-      brackets = 0
-
-      File.each_line(name, chomp: true) do |line|
-        code = line.strip
-
-        if code.ends_with?(/\(|\{|\[/)
-          brackets += code.count("({[")
-          brackets -= code.count(")}]")
-        end
-
-        if !docstring && code.starts_with?("\"\"\"")
-          if under_def || hits.find { |i| i == line_no }
-            docstring = true
-            coverage[line_no] = nil
-            next
-          end
-        end
-
-        # docstring
-        if docstring
-          if code.ends_with?("\"\"\"")
-            docstring = false
-          end
-
-          coverage[line_no] = nil
-          next
-        end
-
-        # comment
-        if code.starts_with?("#")
-          coverage[line_no] = nil
-          next
-        end
-
-        # a hit
-        if hits.find { |i| i == line_no }
-          coverage[line_no] = 1
-          next
-        end
-
-        # inside brackets
-        if brackets > 0
-          coverage[line_no] = nil
-          next
-        end
-
-        # empty string
-        if code.empty?
-          coverage[line_no] = nil
-          next
-        end
-
-        coverage[line_no] = 0
-      ensure
-        line_no += 1
-
-        if code
-          if brackets > 0 && code.ends_with?(/\)|\}|\]/)
-            brackets += code.count("({[")
-            brackets -= code.count(")}]")
-          end
-
-          under_def = code.starts_with?("def ") || code.starts_with?("class ")
-        end
+    def parse(filename : String, error : Process::Stdio = IO::Memory.new) : Array(FileReport)
+      tmpfile = File.tempfile("coverage.xml")
```

Comment:
What if we used [`Tempfile`](https://crystal-lang.org/api/0.23.0/Tempfile.html) with a block, would that solve some of the `ensure` madness?

Reply:
Tempfile does not exist in the latest version.

Reply:
Oops.

Oh but [a `File.tempfile` that takes a block](https://crystal-lang.org/api/1.11.2/File.html#tempfile%28prefix%3AString%7CNil%2Csuffix%3AString%7CNil%2C%2A%2Cdir%3AString%3DDir.tempdir%2Cencoding%3Dnil%2Cinvalid%3Dnil%2C%26%29-class-method) does exist!

Reply:
Maybe we should handle the `File::Error` exception manually, so we're not chasing after errors? We don't care what it does so long as it tries.

```crystal
ensure
  begin
    tmpfile && tmpfile.delete
  rescue File::Error
  end
end
```